fabfile/tasks/helpers.py

```diff
@@ -806,6 +806,38 @@ def setup_hugepages_node(*args):
 def setup_hugepages():
     setup_hugepages_node(env.host_string)
 
+@task
+def setup_coremask_node(*args):
+    """Setup core mask on one or list of nodes
+    USAGE: fab setup_coremask_node:user@host1,user@host2,...
+    """
+    vrouter_file = '/etc/contrail/supervisord_vrouter_files/contrail-vrouter-dpdk.ini'
+    taskset_param = ''
+
+    for host_string in args:
+        dpdk = getattr(env, 'dpdk', None)
+        if dpdk:
+            # do not crash when coremask was not specified in testbed
+            try:
+                coremask = dpdk[env.host_string]['coremask']
+            except KeyError:
+                return
+        else:
+            return
+
+        if (coremask == ''):
+            return
+
+        # supported coremask format: hex: (0x3f); list: (0,3-5), (0,1,2,3,4,5)
+        # if specified as a list of cpus, -c flag must be provided to taskset
+        if ',' in coremask or '-' in coremask:
+            taskset_param = ' -c'
+
+        # prepend taskset only if valid coremask was specified in testbed
+        with settings(host_string=host_string):
```

Comment:
The settings() context manager temporarily overrides env variable passed as argument. In this case we override env.host_string (the left hand operand) with host_string (the right hand operand), which is iterated through args in the for loop (line 817). I agree that a statement like host_string=host_string seems a little weird, but this is convention used throughout the file (lines 29, 624, 737, 776) , so I mimicked the style.

More info on settings() and host_string:
http://docs.fabfile.org/en/latest/usage/env.html#host-string
http://docs.fabfile.org/en/latest/api/core/context_managers.html